include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -245,43 +256,49 @@ T reflectRefract_computeNdotT(bool backside, T NdotI2, T rcpOrientedEta2)
     return r.computeNdotT();
 }
 
-template<typename T NBL_FUNC_REQUIRES(is_scalar_v<T>)
-vector<T,3> reflectRefract_impl(bool _refract, vector<T,3> _I, vector<T,3> _N, T _NdotI, T _NdotTorR, T _rcpOrientedEta)
+template<typename T NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+T reflectRefract_impl(bool _refract, T _I, T _N,
+    typename vector_traits<T>::scalar_type _NdotI,
+    typename vector_traits<T>::scalar_type _NdotTorR,
+    typename vector_traits<T>::scalar_type _rcpOrientedEta)
 {
     return impl::refract<T>::doReflectRefract(_refract, _I, _N, _NdotI, _NdotTorR, _rcpOrientedEta);
 }
 
-template<typename T NBL_FUNC_REQUIRES(is_scalar_v<T>)
-vector<T,3> reflectRefract(bool _refract, vector<T,3> I, vector<T,3> N, bool backside, T NdotI, T NdotI2, T rcpOrientedEta, T rcpOrientedEta2)
+template<typename T NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+T reflectRefract(bool _refract, T I, T N, bool backside, 
+    typename vector_traits<T>::scalar_type NdotI,
+    typename vector_traits<T>::scalar_type NdotI2,
+    typename vector_traits<T>::scalar_type rcpOrientedEta,
+    typename vector_traits<T>::scalar_type rcpOrientedEta2)
 {
     impl::refract<T> r = impl::refract<T>::create(I, N, backside, NdotI, NdotI2, rcpOrientedEta, rcpOrientedEta2);
     return r.doReflectRefract(_refract);
 }
 
-template<typename T NBL_FUNC_REQUIRES(is_scalar_v<T>)
-vector<T,3> reflectRefract(bool _refract, vector<T,3> I, vector<T,3> N, T NdotI, T eta)
+template<typename T NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+T reflectRefract(bool _refract, T I, T N, typename vector_traits<T>::scalar_type NdotI, typename vector_traits<T>::scalar_type eta)
 {
     impl::refract<T> r = impl::refract<T>::create(I, N, NdotI, eta);
     return r.doReflectRefract(_refract);
 }
 
-
 // valid only for `theta` in [-PI,PI]
 template <typename T NBL_FUNC_REQUIRES(is_scalar_v<T>)
 void sincos(T theta, NBL_REF_ARG(T) s, NBL_REF_ARG(T) c)
 {
     c = cos<T>(theta);
-    s = sqrt<T>(1.0-c*c);
-    s = (theta < 0.0) ? -s : s; // TODO: test with XOR
+    s = sqrt<T>(NBL_FP64_LITERAL(1.0)-c*c);
```

Comment:
you need to double bag, `T(NBL_FP64_LITERAL(1.0))` because there's some non-constant foldable subraction before the truncation to fp32